### ISSUE TITLE
Feature Request: ability to replace previous window manager

### ISSUE THREAD

**Issue by dario23** [herbstluftwm/herbstluftwm#838]
some window managers (i verified for openbox) have a mechanism to replace the previously running window manager using some kind of signalling to the previous window manager. i think the code doing this in openbox is in the [`replace_wm` function](https://github.com/danakj/openbox/blob/master/openbox/screen.c#L90).

i'd be happy to look into this more and/or open a PR if there's nothing in principle against adding such functionality.

**Comment by t-wissmann** [comment]
In principle such a feature is a nice to have for hlwm (and if we implement it we should also be so kind and implement the other part of being replaced by another WM).
I was always wondering how this WM replacement works, and I'm wondering whether this works at all when the WM is the magic process of the X-Session (when the magic process terminals, the X-Session is terminated).

For the scenario of WM replacement while the WM is the magic process, herbstluftwm already provides the `wmexec` command that allows to exec into another window manager (openbox supports this too)

**Comment by t-wissmann** [comment]
Two other points:
1. I'm wondering whether there is a protocol for this WM replacement process
2. If the WM is not the magic process, then you can simply kill it and start hlwm...

**Comment by dario23** [comment]
the [change introducing this functionality into xfwm4](https://git.xfce.org/xfce/xfwm4/commit/?h=xfce-4.4&id=4ed599c097aad946513f46b670b966dce2c1c0e2) mentions ICCCM2 compliance.

roughly that seems to say that

  * window managers should acquire the `WM_S<n>` selection ([sec. 4.3](https://tronche.com/gui/x/icccm/sec-4.html#s-4.3))
  * the protocol around transfering ownership is described in [sec. 2](https://tronche.com/gui/x/icccm/sec-2.html#s-2)

**Comment by hannahswain** [comment]
It would be wonderful to see this happen! This is the recommended command in XFCE's [Other Window Manager](https://wiki.xfce.org/howto/other_window_manager) docs.

**Comment by dario23** [comment]
i am not looking into this actively anymore, since for my use-case (testing if things work in another WM quickly/doing one or two things that work better with openbox) is also possible using [openbox's equivalent to the `wmexec` call](http://openbox.org/wiki/Help:Actions#Restart), i.e. the restart action allows you to specify another executable to restart into.

i still think this would be worthwile to have, as not all other WMs have a wmexec command, but some of them still support the more-or-less standardized `--replace` protocol.

**Comment by hannahswain** [comment]
> i still think this would be worthwile to have, as not all other WMs have a wmexec command, but some of them still support the more-or-less standardized --replace protocol.

Indeed, it would be nice to more easily switch. I have HL working as the WM in XFCE using this command from the XFCE documentation: `xfconf-query -c xfce4-session -p /sessions/Failsafe/Client0_Command -n -a -t string -s herbstluftwm`. So there are workarounds, this is just a nice to have.

**Comment by t-wissmann** [comment]
I'm planning to implement this, so the only question is 'when' and not 'if'. Also I try to implement things quicker if more users request it, so thanks for stating that you're also interested in this!

**Comment by t-wissmann** [comment]
Though `--replace` is implemented, it may not fully work because there are two things left: 1. desktop windows seem to be handled wrong if they exist before hlwm, 2. hlwm does not yet connect to the (xfce4-)session manager.
